File: python/00213_house_robber_II.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums[0], nums[1])

        def rob_linear(houses: List[int]) -> int:
            prev1, prev2 = 0, 0
            for amount in houses:
                temp = prev1
                prev1 = max(prev2 + amount, prev1)
                prev2 = temp
            return prev1

        return max(rob_linear(nums[:-1]), rob_linear(nums[1:]))

# Time Complexity: O(n)
# Space Complexity: O(1)

File: java/00213_house_robber_II.java
public class Solution {
    public int rob(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        if (nums.length == 1) {
            return nums[0];
        }
        if (nums.length == 2) {
            return Math.max(nums[0], nums[1]);
        }

        // Compute max between robbing [0..n-2] and [1..n-1]
        return Math.max(rob(nums, 0, nums.length - 2),
                rob(nums, 1, nums.length - 1));
    }

    private int rob(int[] nums, int start, int end) {
        int prev1 = 0, prev2 = 0;

        for (int i = start; i <= end; i++) {
            int temp = prev1;
            prev1 = Math.max(prev2 + nums[i], prev1);
            prev2 = temp;
        }

        return prev1;
    }
}

// Time Complexity: O(n)
// Space Complexity: O(1)
